// @azure/arm-iothub/esm/operations/resourceProviderCommon.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/resourceProviderCommonMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ResourceProviderCommon. */
var ResourceProviderCommon = /** @class */ (function () {
    /**
     * Create a ResourceProviderCommon.
     * @param {IotHubClientContext} client Reference to the service client.
     */
    function ResourceProviderCommon(client) {
        this.client = client;
    }
    ResourceProviderCommon.prototype.getSubscriptionQuota = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, getSubscriptionQuotaOperationSpec, callback);
    };
    return ResourceProviderCommon;
}());
export { ResourceProviderCommon };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getSubscriptionQuotaOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/providers/Microsoft.Devices/usages",
    urlParameters: [
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.UserSubscriptionQuotaListResult
        },
        default: {
            bodyMapper: Mappers.ErrorDetails
        }
    },
    serializer: serializer
};
